File: app/src/main/java/ua/te/hackathon/smartcity2015/ui/main/events/browse/BrowseEventsPresenter.java
package ua.te.hackathon.smartcity2015.ui.main.events.browse;

import android.content.Context;
import android.support.annotation.Nullable;

import ua.te.hackathon.smartcity2015.ui.Presenter;

/**
 * @author victor
 * @since 2016-02-13
 */
public class BrowseEventsPresenter implements Presenter<BrowseEventsView> {
  @Nullable
  private BrowseEventsView view;

  private Context appContext;

  public BrowseEventsPresenter(Context appContext) {
    this.appContext = appContext;
  }

  public void loadEvents() {
    if (view != null) {
      view.showLoadingView();
    }

    // loading

    // loading done
    if (view != null) {
      view.deliverEventList(null);
      view.hideLoadingView();
    }
  }

  @Override
  public void attachView(BrowseEventsView view) {
    this.view = view;
  }

  @Override
  public void detachView() {
    this.view = null;
  }

  @Override
  public void onDestroy() {
    detachView();
  }
}

File: app/src/main/java/ua/te/hackathon/smartcity2015/ui/main/events/browse/BrowseEventsPresenter.kt
package ua.te.hackathon.smartcity2015.ui.main.events.browse

import android.content.Context
import android.location.Location
import io.realm.Realm
import pl.charmas.android.reactivelocation.ReactiveLocationProvider
import ua.te.hackathon.smartcity2015.db.model.Event
import ua.te.hackathon.smartcity2015.sync.SyncManager
import ua.te.hackathon.smartcity2015.ui.base.mvp.Presenter
import ua.te.hackathon.smartcity2015.utils.Logger

/**
 * @author victor
 * *
 * @since 2016-02-13
 */
class BrowseEventsPresenter(private val appContext: Context) : Presenter<BrowseEventsView> {

  companion object Factory {
    val LOG_TAG = Logger.getLogTag(BrowseEventsPresenter::class.java)
  }

  private var view: BrowseEventsView? = null

  private var eventList: List<Event>? = null

  private fun loadLastKnownLocation() {
    val locationProvider = ReactiveLocationProvider(appContext)
    locationProvider.lastKnownLocation.subscribe(
        { location -> findEventsNearBy(location) },
        { error -> onLocationLoadFailed(error) })
  }

  private fun findEventsNearBy(location: Location) {
    Logger.d(LOG_TAG, "Location %f %f".format(location.longitude, location.latitude));

    val realm = Realm.getInstance(appContext)
    val results = realm.allObjects(Event::class.java)

    view?.deliverEventList(results)
    view?.hideLoadingView()
  }

  fun onRefresh() {
    SyncManager.syncUpcomingEvents(appContext)
  }

  private fun onLocationLoadFailed(error: Throwable) {
    Logger.e(LOG_TAG, error)

    view?.deliverLoadingError(error.message)
  }

  fun loadEvents() {
    view?.showLoadingView()

    loadLastKnownLocation()
  }

  override fun attachView(view: BrowseEventsView) {
    this.view = view

    if (eventList == null) {
      loadEvents()
    } else {
      view.deliverEventList(eventList!!)
    }
  }

  override fun detachView() {
    this.view = null
  }

  override fun onDestroy() {
    detachView()
  }
}
